[Messenger] Move Transports to separate packages

Nyholm opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | no
| Deprecations? | no
| Tickets       | 
| License       | MIT
| Doc PR        | coming

I think it is a good idea to have the transports in a separate package. The benefits a many: 
- It allows us to have usage statistics
- The core messenger package is smaller
- Transports can have dependencies specified in composer.json instead of just suggests

This PR will not break BC but it requires to configure subtree split.

Nyholm committed:
[Messenger] Move Transports to separate packages

OskarStark commented on src/Symfony/Component/Messenger/composer.json:
```diff
@@ -17,12 +17,13 @@
         "php": "^7.2.5",
-        "psr/log": "~1.0"
+        "psr/log": "~1.0",
+        "symfony/messenger-amqp": "5.1",
```
Why not ^5.1 ?

Nyholm replied:
Sorry, my mistake.

Nyholm replied:
Thank you

fabpot commented:
I would do it the same way as with the Mailer and Notifier in terms of namespaces. That would mean creating aliases between the old and the new namespaces, but I think it's worth.

jderusse commented on src/Symfony/Component/Messenger/composer.json:
```diff
@@ -17,12 +17,13 @@
+        "psr/log": "~1.0",
+        "symfony/messenger-amqp": "^5.1",
+        "symfony/messenger-doctrine": "^5.1",
+        "symfony/messenger-redis": "^5.1"
```
Should'nt we trigger a deprecation somehow telling `the "XXX" transport has been moved in the package "symfony/messenger-xxx" and will not be included by default in 6.0, run composer require "symfony/messenger-xxx"`?

Nyholm commented:
I've updated the PR: 

* I moved the 3 transports to `Symfony\Component\Messenger\Bridge`. 
* I added class aliases
* I added deprecation notices

Tests passes on all 4 packages. I've been careful and I think I've done everything correct.

jderusse commented on src/Symfony/Component/Messenger/Transport/AmqpExt/AmqpFactory.php:
```diff
@@ -11,25 +11,19 @@
-        return new \AMQPChannel($connection);
-    }
+if (! class_exists(BridgeAmqpFactory::class, false)) {
+    @trigger_error(sprintf('The "%s" class is @deprecated since Symfony 5.1, use "%s" instead. The AmqpExt transport has been moved to package "symfony/messenger-amqp" and will not be required by default in 6.0. Run "composer require symfony/messenger-amqp".', AmqpFactory::class, BridgeAmqpFactory::class), E_USER_DEPRECATED);
```
I don't think we use a `@` before deprecated

Maybe it's just me, but the sentense `will not be required by default` may be understood like "the pakage will not be necessary" which is the oposite. 
What's about `will not be provided in 6.0` or `included`

Nyholm replied:
Oops, thank you. 

I think "included" sounds better. I used "required" because it is composer terminology.

ro0NL commented on src/Symfony/Component/Messenger/composer.json:
```diff
@@ -17,12 +17,13 @@
+        "psr/log": "~1.0",
+        "symfony/messenger-amqp": "^5.1",
+        "symfony/messenger-doctrine": "^5.1",
+        "symfony/messenger-redis": "^5.1"
```
these requires kinda defeat the need for separate packages IMHO.

We should not add them IMHO, as we can never remove them because of BC.

Alternatively keep the legacy transport in place, and let the new bridges extend from those meanwhile.

Nyholm replied:
We can (and will) remove them in 6.0.

Nyholm replied:
These transports should have been in Bridge namespace in the first place. I moving these now, since we are about to add more transports.

ro0NL replied:
so i can require symfony/messenger, while using this new symfony/messenger-amq transport, then updating to 6.0 my transport is removed .. and it crashes?

ro0NL replied:
see https://github.com/symfony/symfony/issues/29229 which triggered me. Maybe we can and should do it in this case.

Nyholm replied:
You are correct. 

Assuming you ignored deprecation messages, the changelog and the UPGRADE docs.

(I just added upgrade docs)

ro0NL replied:
im considering new users, they require symfony/messenger tomorrow, start using the new transports wihout any hint they actually should require it separately.

if we can remove those requires in 6.0 then it's fine, but im worried in Nov 2021 we'll consider it a BC break. In this case, not adding them now would avoid that.

weaverryan replied:
If you require `symfony/messenger` tomorrow and start using the transports, you will get a deprecation warning. The deprecation warnings are actually really nice:

> The "Symfony\Component\Messenger\Transport\Doctrine\DoctrineTransportFactory" class is deprecated since Symfony 5.1, use "Symfony\Component\Messenger\Bridge\Doctrine\Transport\DoctrineTransportFactory" instead. The Doctrine transport has been moved to package "symfony/messenger-doctrine" and will not be included by default in 6.0. Run "composer require symfony/messenger-doctrine".

So the deprecation layer is there - and (unless you ignore the deprecation warnings) you'll know to require the individual packages.

jderusse replied:
are you sure @weaverryan? It triggers the deprecation **only** if you load the legacy class. Which is not the case for a simple usage when you let the Bundle loading the transport (see changes in src/Symfony/Bundle/FrameworkBundle/Resources/config/messenger.xml).

For a migration path shouldn't we:
- let code in `symfony/messenger`
- create a class in `symfony/bridge-transport` that reuse the code (trait or extends) (not an issue as the bridge have a already requires `symfony/messenger`)
- update the bundle to use `symfony/bridge-transport` class when exist, and fallback to `symfony/messenger`
- trigger a deprecation when people use class from `symfony/messenger`

weaverryan reviewed (approved):
I just played with this code using our (fairly advanced) SfCasts Messenger tutorial and it worked flawlessly. The deprecation warnings are really nice - each instructing me which package to require to fix the deprecation. I can't test requiring the new packages yet (since they don't exist), but that's easy to try after merge and creating the sub-tree splits.

One missing piece I discussed with Tobias (and he agreed, but for a different PR) is to add code to *suggest* (with an exception) that you require a specific package if you try to use a "core" transport. We should add that "now" (but in a future PR), though users won't see this message until 6.0 when they try to use a transport that's not installed.

+1 from me - I can't find any issues.

nicolas-grekas commented on src/Symfony/Component/Messenger/Bridge/AmqpExt/CHANGELOG.md:
```diff
@@ -0,0 +1,7 @@
+5.1.0
+-----
+
+ * Moved the transport to a separate package.
```
`Introduced the bridge`, no?

Nyholm replied:
Thank you

nicolas-grekas commented on UPGRADE-6.0.md:
```diff
@@ -19,6 +19,13 @@ HttpFoundation
+Messenger
+---------
+
+ * Removed AmqpExt transport. Run `composer require symfony/messenger-amqp` to keep the transport in your application..
```
double end dot

Nyholm replied:
Thank you

Nyholm commented:
I moved `class_alias` to the new class after discussions with Nicolas. 

The following does not work: 

```php
// FILE: old.php
namespace Old;
class_alias(\New\Foo::class, \Old\Foo::class);

if (false) {
    class Foo{}
}

// FILE: new.php
namespace New;
class Foo{}

// FILE: index.php
function foo(\Old\Foo $old) {};

// This will break. 
foo(new \New\Foo());
```

We need to move the `class_alias` to the new.php. 

```php
// FILE: old.php
namespace Old;
// Load the new class
class_exists(\New\Foo::class);

if (false) {
    class Foo{}
}

// FILE: new.php
namespace New;
class Foo{}
class_alias(Foo::class, Old\Foo::class);

// FILE: index.php
function foo(\Old\Foo $old) {};

// This will work
foo(new {New\Foo());
```

nicolas-grekas reviewed (approved):
The BC layer looks good enough for me.
Thank you.

jderusse commented on src/Symfony/Component/Messenger/Bridge/AmqpExt/composer.json:
```diff
@@ -0,0 +1,41 @@
+    ],
+    "require": {
+        "php": "^7.2.5",
+        "ext-amqp": "*",
```
I get why the ext is required here. But the extension `amsp` wasn't required in `symfony/messenger`. Given the bridge is a requirement for messenger, now `symfony/messenger` requires amqp+redis+doctrine, which may not be really needed

same for ext-redis.

Nyholm replied:
Wow. You are correct!

Great catch. 
They should be added in 6.0. 

I’m on my phone now. Will update later tonight

Nyholm replied:
This is resolved.

Nyholm commented:
Thank you for all the reviews and carefully thinking out the BC layer.

jderusse commented on src/Symfony/Component/Messenger/Bridge/AmqpExt/composer.json:
```diff
@@ -0,0 +1,40 @@
+{
+    "name": "symfony/messenger-amqp",
```
shouldn't it be `symfony/amqp-messenger` for consistency with `symfony/slack-notifier` and `symfony/mailchimp-mailer`?

Nyholm replied:
Thank you.

Tobion commented on src/Symfony/Bundle/FrameworkBundle/Resources/config/messenger.xml:
```diff
@@ -67,11 +67,11 @@
         </service>
 
-        <service id="messenger.transport.amqp.factory" class="Symfony\Component\Messenger\Transport\AmqpExt\AmqpTransportFactory">
+        <service id="messenger.transport.amqp.factory" class="Symfony\Component\Messenger\Bridge\AmqpExt\Transport\AmqpTransportFactory">
```
What happens when you configure an amqp transport but do not have the new bridge installed?

Line 51, still references the old `RejectRedeliveredMessageMiddleware`

Nyholm replied:
We are removing them in the FrameworkExtension

Tobion commented on src/Symfony/Component/Messenger/Bridge/AmqpExt/Middleware/RejectRedeliveredMessageMiddleware.php:
```diff
@@ -0,0 +1,53 @@
+    {
+        $amqpReceivedStamp = $envelope->last(AmqpReceivedStamp::class);
+        if ($amqpReceivedStamp instanceof AmqpReceivedStamp && $amqpReceivedStamp->getAmqpEnvelope()->isRedelivery()) {
+            throw new RejectRedeliveredMessageException('Redelivered message from AMQP detected that will be rejected and trigger the retry logic.');
```
The original idea was that the RejectRedeliveredMessageMiddleware can be implemented also for other transports like doctrine. So not necessarily specific to amqp. But I guess we can then also just add a new middleware in in the doctrine bridge if needed at some point.
So it's fine to move this to the bridge. But then I would also suggest to move the `RejectRedeliveredMessageException` to the amqp bridge.

Tobion replied:
Another thing is that we would need to conditionally enable this middleware when the bridge is installed. Currently it's always active but in 6.0 it would just error when the amqp bridge is not installed.

Nyholm replied:
I moved back the middleware to the core package. I did it for the following reasons: 
1) It was the original idea to let other transports use this middleware
2) I cannot see an easy way to conditionally add this as default middleware base on if it is installed or not.

Someone may decide to move it to the bridge in the future. But Im leaving it here for now.

Tobion replied:
> I cannot see an easy way to conditionally add this as default middleware base on if it is installed or not.

Isn't that as simple as wrapping https://github.com/symfony/symfony/blob/master/src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php#L1584 around a class_exists call?

Nyholm replied:
There are two classes you need to check. You cannot simply wrap it, you need to write something like this: 

```php
$before[] = ['id' => 'add_bus_name_stamp_middleware'];
if (class_exists(LegacyMiddleware) || class_exists(BridgeyMiddleware)) {
    $before[] = ['id' => 'reject_redelivered_message_middleware'];
}
$before[] = ['id' => 'dispatch_after_current_bus'];
$before[] = ['id' => 'failed_message_processing_middleware'];

$defaultMiddleware = [
    'before' => $before,
    'after' => [
        ['id' => 'send_message'],
        ['id' => 'handle_message'],
    ],
];
```

I didnt feel like that was a very pretty solution (Maybe the "prettiest", but not "pretty"). So I decided to not move it, since you said that the idea was to let other transports add their logic here.

Nyholm commented:
Thank you @Tobion for your review.

Tobion commented:
The transports are currently inside the bridge folder of the symfony/messenger package. But for 6.0 we don't want to include the transports automatically. But I don't see how that is going to work without moving the files again? Shouldn't we create new packages outside the symfony/messenger for this to be equipped for 6.0?

Nyholm commented:
For what I understand, we can configure subtreesplit so it works as expected. Just like with the Notifier and Mailer component. 

So we dont need to move the files again.

Tobion commented:
I see. If it works like in mailer that's fine for me. I just think it's not an ideal setup because people will not realize that the Bridges are not part of symfony/messenger from reading the source of symfony/symfony on Github.

fabpot reviewed (commented):
I'm not a big fan of the names: why not `Amqp` instead of `AmqpExt`? Same for Redis. The DSN use `redis` and `amqp` anyway, so I don't see a good reason to use `Ext` in the namespace (and IIRC, we are doing it anywhere else).

Nyholm commented:
Sure. I'll rename it. If we want to support the amqp-lib in a separate transport, it could be named `AmqpLib`.

Nyholm commented:
The namespace is renamed, and the tests are green.

fabpot commented:
Thank you @Nyholm.

Nyholm commented:
Thank you for merging and thank you for all reviews.

Tobion commented:
This broke BC for the stamps, see https://github.com/symfony/symfony/issues/37616

stof commented on UPGRADE-5.1.md:
```diff
@@ -19,6 +19,13 @@ HttpFoundation
+---------
+
+ * Deprecated AmqpExt transport. It has moved to a separate package. Run `composer require symfony/amqp-messenger` to use the new classes.
+ * Deprecated Doctrine transport. It has moved to a separate package. Run `composer require symfony/doctrine-messenger` to use the new classes.
```
some nitpick. I would rename this one to `doctrine-dbal-messenger`. The Doctrine organization has many separate projects related to persistence. This bridge integrates with Doctrine DBAL, not with the whole Doctrine projects (having a single bridge for the whole Doctrine makes the dependency management a pain for `symfony/doctrine-bridge` btw, so if we ever have another transport using a different Doctrine project, I'd rather have a separate bridge for it)
